src/spark/utils.py

```diff
@@ -3,57 +3,94 @@
-    :param delta_lake: build the spark session with delta lake support
 
-    :return: A `SparkSession` object
+    :return: A SparkConf object with the base configuration
     """
+    return SparkConf().setAll([
+        ("spark.master", os.environ.get("SPARK_MASTER_URL", "spark://spark-master:7077")),
+        ("spark.app.name", app_name),
+    ])
+
 
+def get_spark_session(
+        app_name: str,
+        local: bool = False,
+        delta_lake: bool = False) -> SparkSession:
+    """
+    Helper to get and manage the SparkSession and keep all of our spark configuration params in one place.
+
+    :param app_name: The name of the application
+    :param local: Whether to run the spark session locally or not
+    :param delta_lake: Build the spark session with Delta Lake support
+
+    :return: A SparkSession object
+    """
     if local:
         return SparkSession.builder.appName(app_name).getOrCreate()
 
-    spark_conf = SparkConf()
+    spark_conf = get_base_spark_conf(app_name)
 
     if delta_lake:
 
-        jars_dir = "/opt/bitnami/spark/jars/"
-        jar_files = [os.path.join(jars_dir, f) for f in os.listdir(jars_dir) if f.endswith(".jar")]
-        jars = ",".join(jar_files)
-
-        spark_conf.setAll(
-            [
-                (
-                    "spark.master",
-                    os.environ.get("SPARK_MASTER_URL", "spark://spark-master:7077"),
-                ),
-                ("spark.app.name", app_name),
-                ("spark.hadoop.fs.s3a.endpoint", os.environ.get("MINIO_URL")),
-                ("spark.hadoop.fs.s3a.access.key", os.environ.get("MINIO_ACCESS_KEY")),
-                ("spark.hadoop.fs.s3a.secret.key", os.environ.get("MINIO_SECRET_KEY")),
-                ("spark.jars", jars),
-                ("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension"),
-                ("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog"),
-                ("spark.hadoop.fs.s3a.path.style.access", "true"),
-                ("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem"),
-            ]
-        )
-    else:
-        spark_conf.setAll(
-            [
-                (
-                    "spark.master",
-                    os.environ.get("SPARK_MASTER_URL", "spark://spark-master:7077"),
-                ),
-                ("spark.app.name", app_name),
-            ]
-        )
-
-    return SparkSession.builder.config(conf=spark_conf).getOrCreate()
+        # Just to include the necessary jars for Delta Lake
+        jar_names = [f"delta-spark_{SCALA_VER}-{DELTA_SPARK_VER}.jar",
+                     f"hadoop-aws-{HADOOP_AWS_VER}.jar"]
+        jars_str = _get_jars(jar_names)
+        delta_conf = _get_delta_lake_conf(jars_str)
+        for key, value in delta_conf.items():
+            spark_conf.set(key, value)
+
+    return SparkSession.builder.config(conf=spark_conf).enableHiveSupport().getOrCreate()
```

Comment:
I don't know anything about hive, really - why is it needed here? Doesn't DeltaLake take care of the metadata?

Reply:
with this approach, we can establish a permanent view that everyone can query without rebuilding/reloading the table. I removed `enableHiveSupport()` from general spark session builder and only enable this via conf for delta lake.

Reply:
So the delta lake SW doesn't maintain the table metadata anywhere other than local memory?

Where are the metadata persisted?

Reply:
Yea, in memory by default. The metadata is persisted in the metastore_db folder which is created by default under /cdm_shared_folder and mounted by the rancher settings.

Reply:
So anyone wanting to use the shared tables needs the same volume mount for their notebook container? What happens if two people try to create conflicting tables with the same name at the same time?

Reply:
I believe table is locked for concurrent access. I don’t understand the first question. Does that matter?

Reply:
The table metadata is stored in a file on the volume mount, right? So If you're working in notebook container A and create a table, that metadata gets stored in the volume mount, which is actually some real place outside the container. If I'm working in notebook container B, then I need that same file mounted into my container to see the table you made.

Will the concurrent locks work though the different containers? E.g. if I lock  the files in container B will they be locked for you in container A?

Reply:
Yea the metadata is stored in the shared volume. If you are looking to read metadata the container need to be mounted the same volume which is exactly our current setup. The lock is on the metadata file so it’s locked for all containers.

Reply:
The reason I'm asking is that for some flavors of NFS os level file locks don't actually work and allow writing to the file on other machines while it's locked on the first machine. With how quirky docker can be I'm wondering if it has the same sort of problems

Reply:
yea I notice that behavior. In Jupiter notebook I gave some instructions of how to delete the DB locks manually in the event if accessing issues.

Reply:
What's the lock mechanism?

Reply:
There is a .lck file. Just delete that file.

Reply:
Ok, that seems like that should work regardless of docker

Reply:
Can you check and see if there's anyway for the tables to be stored in minio so other systems that can talk deltalake can access them?

Reply:
yea. I will look into that.